File: Task2/src/py/wup_cca2.py
#!/usr/bin/env python3
"""Task2: WUP protocol + bit-by-bit CCA2 attack (reproducing Knockel et al. 2018 §4.1).

Workflow:
  1. Server generates its own 1024-bit RSA key;
  2. Client generates a 128-bit AES session key, encrypts it with RSA, and uses it to encrypt a WUP request via AES-128-ECB;
  3. Generate historical messages (RSA ciphertext + AES ciphertext);
  4. Attacker recovers the AES key bit-by-bit using only the public key, historical messages, and the server's ACCEPT/REJECT oracle;
  5. Decrypt the historical WUP using the recovered key.

Output (written to Task2/ directory):
  AES_Key.txt / WUP_Request.txt / AES_Encrypted_WUP.txt / History_Message.txt / attack_log.txt
"""

import secrets
import sys
from pathlib import Path

_HERE = Path(__file__).resolve()
sys.path.insert(0, str(_HERE.parents[3] / "Task1" / "src" / "py"))
sys.path.insert(0, str(_HERE.parent))
import rsa_core
import aes_utils
from wup import build_wup_request, valid_wup

TASK2_DIR = _HERE.parents[2]
PROBE_NONCE = "cca2-probe"


def low128_bytes(value: int) -> bytes:
    """Take the lowest 128 bits of the integer, exporting as 16 bytes (big-endian, right-aligned keeping leading zeros)."""
    return (value & ((1 << 128) - 1)).to_bytes(16, "big")


class WUPServer:
    """Holds the RSA private key and exposes only the ACCEPT/REJECT oracle query()->bool.

    Reproduces the key behavior of QQ Browser: after RSA decryption, only the lowest 128 bits are taken as the AES session key.
    """

    def __init__(self, n: int, e: int, d: int):
        self._n = n
        self._e = e
        self._d = d

    @property
    def public_key(self) -> tuple[int, int]:
        return self._n, self._e

    def query(self, rsa_ciphertext: int, aes_ciphertext: bytes) -> bool:
        rsa_plain = rsa_core.rsa_decrypt_int(rsa_ciphertext, self._d, self._n)
        key = low128_bytes(rsa_plain)
        request = aes_utils.aes_ecb_decrypt(key, aes_ciphertext)
        return request is not None and valid_wup(request)


class CCA2Attacker:
    """Only holds the public key and oracle callback, without access to the private key d or the actual AES key."""

    def __init__(self, public_key: tuple[int, int], oracle, victim_rsa_ct: int):
        self._n, self._e = public_key
        self._oracle = oracle
        self._victim_c = victim_rsa_ct
        self._probe = build_wup_request(PROBE_NONCE)

    def recover_key(self, log_lines: list[str]) -> tuple[int, int]:
        known = 0
        query_count = 0
        for bit in range(128):
            shift = 127 - bit
            factor = rsa_core.rsa_encrypt_int(1 << shift, self._e, self._n)
            shifted_rsa = self._victim_c * factor % self._n
            known_before = known

            guess_zero_key = low128_bytes(known << shift)
            encrypted_probe = aes_utils.aes_ecb_encrypt(guess_zero_key, self._probe)

            accepted = self._oracle(shifted_rsa, encrypted_probe)
            query_count += 1
            recovered_bit = 0 if accepted else 1
            if not accepted:
                known |= 1 << bit

            log_lines.append(
                f"[Round {bit + 1:03d}]\n"
                f"target_original_bit={bit}\n"
                f"shift={shift}\n"
                f"transformed_rsa_ciphertext={format(shifted_rsa, 'x')}\n"
                f"known_low_bits_before=0x{format(known_before, 'x')}\n"
                f"candidate_bit=0\n"
                f"candidate_server_key={guess_zero_key.hex()}\n"
                f"oracle_result={'ACCEPT' if accepted else 'REJECT'}\n"
                f"recovered_bit={recovered_bit}\n"
                f"known_low_bits_after=0x{format(known, 'x')}\n"
            )
            if (bit + 1) % 16 == 0:
                print(f"recovered {bit + 1}/128 bits")
        return known, query_count

def main() -> int:
    keys = rsa_core.generate_rsa_keypair(1024)
    n, e, d = keys["n"], keys["e"], keys["d"]
    server = WUPServer(n, e, d)

    # --- Client: Generate session key and historical messages ---
    aes_key = secrets.token_bytes(16)
    plain_request = build_wup_request("20260616")
    rsa_ct = rsa_core.rsa_encrypt_int(rsa_core.bytes_to_int(aes_key), e, n)
    encrypted_request = aes_utils.aes_ecb_encrypt(aes_key, plain_request)

    (TASK2_DIR / "AES_Key.txt").write_text(aes_key.hex() + "\n")
    (TASK2_DIR / "WUP_Request.txt").write_text(plain_request.hex() + "\n")
    (TASK2_DIR / "AES_Encrypted_WUP.txt").write_text(encrypted_request.hex() + "\n")
    (TASK2_DIR / "History_Message.txt").write_text(
        f"rsa_n_hex={format(n, 'x')}\n"
        f"rsa_e_dec={e}\n"
        f"rsa_encrypted_aes_key_hex={format(rsa_ct, 'x')}\n"
        f"aes_encrypted_wup_request_hex={encrypted_request.hex()}\n"
    )
    print("history written to History_Message.txt")
    print(f"RSA-encrypted AES key: {format(rsa_ct, 'x')}")
    print(f"AES-encrypted WUP request: {encrypted_request.hex()}\n")

    # --- Attacker: Recover key using only the public key + oracle ---
    attacker = CCA2Attacker(server.public_key, server.query, rsa_ct)
    log_lines: list[str] = []
    recovered, query_count = attacker.recover_key(log_lines)
    recovered_key = low128_bytes(recovered)

    recovered_request = aes_utils.aes_ecb_decrypt(recovered_key, encrypted_request)
    recovered_plaintext = (
        recovered_request.decode("utf-8", errors="replace")
        if recovered_request is not None else ""
    )
    keys_match = recovered_key == aes_key

    log_lines.append(
        f"\nactual_aes_key={aes_key.hex()}\n"
        f"recovered_aes_key={recovered_key.hex()}\n"
        f"keys_match={'true' if keys_match else 'false'}\n"
        f"history_wup_plaintext={recovered_plaintext}"
        f"query_count={query_count}\n"
    )
    (TASK2_DIR / "attack_log.txt").write_text("\n".join(log_lines))

    print(f"\nactual AES key:    {aes_key.hex()}")
    print(f"recovered AES key: {recovered_key.hex()}")
    print(f"keys_match: {keys_match}")
    print(f"query_count: {query_count}")
    print("decrypted historical request:")
    print(recovered_plaintext)
    return 0 if keys_match else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: Task2/src/cpp/wup_cca2.cpp
#include "../../../Task1/src/cpp/rsa_utils.h"

#include <fstream>
#include <gmpxx.h>
#include <iomanip>
#include <iostream>
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

using Bytes = std::vector<unsigned char>;

Bytes bytes(const std::string &s) { return Bytes(s.begin(), s.end()); }

std::string hex(const Bytes &data) {
	std::ostringstream out;
	for (unsigned char c : data)
		out << std::hex << std::setw(2) << std::setfill('0') << int(c);
	return out.str();
}

mpz_class mpz_from_bytes(const Bytes &data) {
	mpz_class x;
	mpz_import(x.get_mpz_t(), data.size(), 1, 1, 1, 0, data.data());
	return x;
}


Bytes low128(mpz_class x) {
	static const mpz_class mask = (mpz_class(1) << 128) - 1;
	x &= mask;

	Bytes out(16), tmp(16);
	size_t written = 0;
	mpz_export(tmp.data(), &written, 1, 1, 1, 0, x.get_mpz_t());
	std::copy(tmp.begin(), tmp.begin() + written, out.end() - written);
	return out;
}

Bytes random_bytes(size_t n) {
	Bytes out(n);
	if (RAND_bytes(out.data(), out.size()) != 1)
		throw std::runtime_error("RAND_bytes failed");
	return out;
}

Bytes aes_ecb_encrypt(const Bytes &plain, const Bytes &key) {
	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
	Bytes out(plain.size() + 16);
	int len = 0, total = 0;

	EVP_EncryptInit_ex(ctx, EVP_aes_128_ecb(), nullptr, key.data(), nullptr);
	EVP_EncryptUpdate(ctx, out.data(), &len, plain.data(), plain.size());
	total += len;
	EVP_EncryptFinal_ex(ctx, out.data() + total, &len);
	total += len;

	EVP_CIPHER_CTX_free(ctx);
	out.resize(total);
	return out;
}

std::optional<Bytes> aes_ecb_decrypt(const Bytes &cipher, const Bytes &key) {
	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
	Bytes out(cipher.size());
	int len = 0, total = 0;

	EVP_DecryptInit_ex(ctx, EVP_aes_128_ecb(), nullptr, key.data(), nullptr);
	if (EVP_DecryptUpdate(ctx, out.data(), &len, cipher.data(), cipher.size()) !=
		1) {
		EVP_CIPHER_CTX_free(ctx);
		return std::nullopt;
	}
	total += len;
	if (EVP_DecryptFinal_ex(ctx, out.data() + total, &len) != 1) {
		EVP_CIPHER_CTX_free(ctx);
		return std::nullopt;
	}
	total += len;

	EVP_CIPHER_CTX_free(ctx);
	out.resize(total);
	return out;
}

bool valid_wup(const Bytes &plain) {
	const std::string s(plain.begin(), plain.end());
	return s.rfind("WUP/1\n", 0) == 0 && s.find("\ncmd=") != std::string::npos &&
		   s.size() >= 4 && s.substr(s.size() - 4) == "END\n";
}

std::string wup_request(const std::string &nonce) {
	return "WUP/1\ncmd=sync\nuid=10001\nnonce=" + nonce +
		   "\nbody=browser-history-demo\nEND\n";
}

struct Server {
	mpz_class n, e, d;

	std::optional<Bytes> handle(const mpz_class &rsa_key,
								const Bytes &encrypted_request) const {
		const Bytes key = low128(textbookRSA::decrypt(rsa_key, d, n));
		const auto request = aes_ecb_decrypt(encrypted_request, key);
		if (!request || !valid_wup(*request))
			return std::nullopt;
		return aes_ecb_encrypt(bytes("WUP/1\nstatus=ok\nEND\n"), key);
	}
};

struct History {
	Bytes aes_key;
	mpz_class rsa_key;
	Bytes encrypted_request;
	std::string plain_request;
};

History client_send(const Server &server) {
	History h;
	h.aes_key = random_bytes(16);
	h.plain_request = wup_request("20260616");
	h.rsa_key = textbookRSA::encrypt(mpz_from_bytes(h.aes_key), server.e, server.n);
	h.encrypted_request = aes_ecb_encrypt(bytes(h.plain_request), h.aes_key);
	return h;
}

void write_history(const Server &server, const History &h) {
	std::ofstream out(textbookRSA::resolve_key_path("../../History_Message.txt"));
	out << "rsa_n_hex=" << server.n.get_str(16) << "\n";
	out << "rsa_e_dec=" << server.e << "\n";
	out << "rsa_encrypted_aes_key_hex=" << h.rsa_key.get_str(16) << "\n";
	out << "aes_encrypted_wup_request_hex=" << hex(h.encrypted_request) << "\n";
}

// Save output files.
//   AES_Key.txt          : 128-bit AES key, 32 lowercase hex characters, keeping leading zeros
//   WUP_Request.txt      : Hex of raw WUP request bytes
//   AES_Encrypted_WUP.txt: Hex of AES-128-ECB ciphertext
void write_client_files(const History &h) {
	std::ofstream(textbookRSA::resolve_key_path("../../AES_Key.txt"))
		<< hex(h.aes_key) << "\n";
	std::ofstream(textbookRSA::resolve_key_path("../../WUP_Request.txt"))
		<< hex(bytes(h.plain_request)) << "\n";
	std::ofstream(textbookRSA::resolve_key_path("../../AES_Encrypted_WUP.txt"))
		<< hex(h.encrypted_request) << "\n";
}

mpz_class recover_key(const Server &server, const History &h, std::ostream &log,
					  int &query_count) {
	mpz_class known = 0;
	const Bytes probe = bytes(wup_request("cca2-probe"));
	query_count = 0;

	for (int bit = 0; bit < 128; ++bit) {
		const int shift = 127 - bit;
		const mpz_class factor = mpz_class(1) << shift;
		const mpz_class shifted_rsa =
			h.rsa_key * textbookRSA::encrypt(factor, server.e, server.n) % server.n;
		const mpz_class known_before = known;
		const Bytes guess_zero_key = low128(known << shift);
		const Bytes encrypted_probe = aes_ecb_encrypt(probe, guess_zero_key);

		const bool accepted = server.handle(shifted_rsa, encrypted_probe).has_value();
		++query_count;
		const int recovered_bit = accepted ? 0 : 1;
		if (!accepted)
			known |= mpz_class(1) << bit;

		log << "[Round " << std::setw(3) << std::setfill('0') << (bit + 1) << "]\n"
			<< "target_original_bit=" << bit << "\n"
			<< "shift=" << shift << "\n"
			<< "transformed_rsa_ciphertext=" << shifted_rsa.get_str(16) << "\n"
			<< "known_low_bits_before=0x" << known_before.get_str(16) << "\n"
			<< "candidate_bit=0\n"
			<< "candidate_server_key=" << hex(guess_zero_key) << "\n"
			<< "oracle_result=" << (accepted ? "ACCEPT" : "REJECT") << "\n"
			<< "recovered_bit=" << recovered_bit << "\n"
			<< "known_low_bits_after=0x" << known.get_str(16) << "\n\n";

		if ((bit + 1) % 16 == 0)
			std::cout << "recovered " << bit + 1 << "/128 bits\n";
	}
	return known;
}

int main() {
	Server server;
	textbookRSA::generate_keys(1024, server.n, server.e, server.d);

	const History history = client_send(server);
	write_history(server, history);
	write_client_files(history);

	std::cout << "history written to History_Message.txt\n";
	std::cout << "RSA-encrypted AES key: " << history.rsa_key.get_str(16) << "\n";
	std::cout << "AES-encrypted WUP request: " << hex(history.encrypted_request)
			  << "\n\n";

	std::ofstream log(textbookRSA::resolve_key_path("../../attack_log.txt"));
	int query_count = 0;
	const mpz_class recovered = recover_key(server, history, log, query_count);
	const Bytes recovered_key = low128(recovered);
	const auto recovered_request =
		aes_ecb_decrypt(history.encrypted_request, recovered_key);
	const std::string recovered_plaintext =
		recovered_request
			? std::string(recovered_request->begin(), recovered_request->end())
			: std::string();
	const bool keys_match = recovered_key == history.aes_key;

	log << "actual_aes_key=" << hex(history.aes_key) << "\n"
		<< "recovered_aes_key=" << hex(recovered_key) << "\n"
		<< "keys_match=" << (keys_match ? "true" : "false") << "\n"
		<< "history_wup_plaintext=" << recovered_plaintext << "\n"
		<< "query_count=" << query_count << "\n";

	std::cout << "\nactual AES key:    " << hex(history.aes_key) << "\n";
	std::cout << "recovered AES key: " << hex(recovered_key) << "\n";
	std::cout << "decrypted historical request:\n" << recovered_plaintext;
	return 0;
}
